File: app/haskell/part3/test/CompositePatternSpec.hs
module CompositePatternSpec (spec) where

import Test.Hspec
import Test.QuickCheck
import qualified Data.Map.Strict as Map

import CompositePattern

spec :: Spec
spec = do
  shapeSpec
  switchableSpec
  fileSystemSpec
  menuItemSpec
  expressionSpec
  organizationSpec

-- ============================================================
-- Shape Tests
-- ============================================================

shapeSpec :: Spec
shapeSpec = describe "Shape" $ do
  describe "Circle" $ do
    it "should translate correctly" $ do
      let circle = Circle (Point 10 10) 5
          translated = translateShape 5 5 circle
      translated `shouldBe` Circle (Point 15 15) 5

    it "should scale correctly" $ do
      let circle = Circle (Point 10 10) 5
          scaled = scaleShape 2 circle
      scaled `shouldBe` Circle (Point 10 10) 10

    it "should calculate area correctly" $ do
      let circle = Circle (Point 0 0) 5
      shapeArea circle `shouldBe` pi * 25

  describe "Square" $ do
    it "should translate correctly" $ do
      let square = Square (Point 0 0) 10
          translated = translateShape 3 4 square
      translated `shouldBe` Square (Point 3 4) 10

    it "should scale correctly" $ do
      let square = Square (Point 0 0) 10
          scaled = scaleShape 2 square
      scaled `shouldBe` Square (Point 0 0) 20

    it "should calculate area correctly" $ do
      let square = Square (Point 0 0) 10
      shapeArea square `shouldBe` 100

  describe "Rectangle" $ do
    it "should translate correctly" $ do
      let rect = Rectangle (Point 0 0) 10 5
          translated = translateShape 2 3 rect
      translated `shouldBe` Rectangle (Point 2 3) 10 5

    it "should scale correctly" $ do
      let rect = Rectangle (Point 0 0) 10 5
          scaled = scaleShape 2 rect
      scaled `shouldBe` Rectangle (Point 0 0) 20 10

    it "should calculate area correctly" $ do
      let rect = Rectangle (Point 0 0) 10 5
      shapeArea rect `shouldBe` 50

  describe "CompositeShape" $ do
    it "should create empty composite" $ do
      emptyComposite `shouldBe` CompositeShape []

    it "should add shapes to composite" $ do
      let circle = Circle (Point 0 0) 5
          square = Square (Point 10 10) 10
          composite = addShape square (addShape circle emptyComposite)
      composite `shouldBe` CompositeShape [circle, square]

    it "should translate all children" $ do
      let circle = Circle (Point 0 0) 5
          square = Square (Point 10 10) 10
          composite = addShape square (addShape circle emptyComposite)
          translated = translateShape 5 5 composite
      translated `shouldBe` CompositeShape 
        [ Circle (Point 5 5) 5
        , Square (Point 15 15) 10
        ]

    it "should scale all children" $ do
      let circle = Circle (Point 0 0) 5
          square = Square (Point 10 10) 10
          composite = addShape square (addShape circle emptyComposite)
          scaled = scaleShape 2 composite
      scaled `shouldBe` CompositeShape 
        [ Circle (Point 0 0) 10
        , Square (Point 10 10) 20
        ]

    it "should calculate total area" $ do
      let circle = Circle (Point 0 0) 5  -- area = pi * 25
          square = Square (Point 0 0) 10  -- area = 100
          composite = addShape square (addShape circle emptyComposite)
      shapeArea composite `shouldBe` (pi * 25 + 100)

    it "should support nested composites" $ do
      let circle = Circle (Point 0 0) 5
          square = Square (Point 0 0) 10
          inner = addShape square (addShape circle emptyComposite)
          outer = addShape inner emptyComposite
          translated = translateShape 1 1 outer
      case translated of
        CompositeShape [CompositeShape [Circle p1 _, Square p2 _]] -> do
          p1 `shouldBe` Point 1 1
          p2 `shouldBe` Point 1 1
        _ -> expectationFailure "Unexpected structure"

-- ============================================================
-- Switchable Tests
-- ============================================================

switchableSpec :: Spec
switchableSpec = describe "Switchable" $ do
  describe "Light" $ do
    it "should turn on" $ do
      let light = Light "Ceiling" False
          onLight = turnOn light
      onLight `shouldBe` Light "Ceiling" True
      isOn onLight `shouldBe` True

    it "should turn off" $ do
      let light = Light "Ceiling" True
          offLight = turnOff light
      offLight `shouldBe` Light "Ceiling" False
      isOn offLight `shouldBe` False

  describe "DimmableLight" $ do
    it "should turn on to max intensity" $ do
      let light = DimmableLight "Bedside" 0
          onLight = turnOn light
      onLight `shouldBe` DimmableLight "Bedside" 100
      isOn onLight `shouldBe` True

    it "should turn off to zero intensity" $ do
      let light = DimmableLight "Bedside" 50
          offLight = turnOff light
      offLight `shouldBe` DimmableLight "Bedside" 0
      isOn offLight `shouldBe` False

  describe "Fan" $ do
    it "should turn on" $ do
      let fan = Fan "CeilingFan" False 0
          onFan = turnOn fan
      onFan `shouldBe` Fan "CeilingFan" True 1
      isOn onFan `shouldBe` True

    it "should turn off but preserve speed" $ do
      let fan = Fan "CeilingFan" True 3
          offFan = turnOff fan
      offFan `shouldBe` Fan "CeilingFan" False 3
      isOn offFan `shouldBe` False

  describe "CompositeSwitchable" $ do
    it "should create empty composite" $ do
      let composite = emptySwitchableComposite "Room"
      composite `shouldBe` CompositeSwitchable "Room" []

    it "should add switchables" $ do
      let light = Light "Ceiling" False
          fan = Fan "Fan" False 0
          composite = addSwitchable fan (addSwitchable light (emptySwitchableComposite "Room"))
      composite `shouldBe` CompositeSwitchable "Room" [light, fan]

    it "should turn on all children" $ do
      let light = Light "Ceiling" False
          dimmable = DimmableLight "Lamp" 0
          composite = addSwitchable dimmable (addSwitchable light (emptySwitchableComposite "Room"))
          onComposite = turnOn composite
      onComposite `shouldBe` CompositeSwitchable "Room" 
        [ Light "Ceiling" True
        , DimmableLight "Lamp" 100
        ]
      isOn onComposite `shouldBe` True

    it "should turn off all children" $ do
      let light = Light "Ceiling" True
          dimmable = DimmableLight "Lamp" 100
          composite = addSwitchable dimmable (addSwitchable light (emptySwitchableComposite "Room"))
          offComposite = turnOff composite
      offComposite `shouldBe` CompositeSwitchable "Room" 
        [ Light "Ceiling" False
        , DimmableLight "Lamp" 0
        ]
      isOn offComposite `shouldBe` False

    it "should report isOn if any child is on" $ do
      let light = Light "Ceiling" False
          dimmable = DimmableLight "Lamp" 50
          composite = addSwitchable dimmable (addSwitchable light (emptySwitchableComposite "Room"))
      isOn composite `shouldBe` True

-- ============================================================
-- FileSystem Tests
-- ============================================================

fileSystemSpec :: Spec
fileSystemSpec = describe "FileSystem" $ do
  describe "File" $ do
    it "should have correct size" $ do
      let file = File "readme.txt" 1024 ".txt"
      fileSize file `shouldBe` 1024

    it "should count as 1 file" $ do
      let file = File "readme.txt" 1024 ".txt"
      fileCount file `shouldBe` 1

  describe "Directory" $ do
    it "should calculate total size" $ do
      let readme = File "readme.txt" 1024 ".txt"
          config = File "config.json" 512 ".json"
          dir = Directory "project" [readme, config]
      fileSize dir `shouldBe` 1536

    it "should count total files" $ do
      let readme = File "readme.txt" 1024 ".txt"
          config = File "config.json" 512 ".json"
          src = Directory "src" 
            [ File "main.fs" 2048 ".fs"
            , File "util.fs" 1024 ".fs"
            ]
          root = Directory "project" [readme, config, src]
      fileCount root `shouldBe` 4

    it "should find files by extension" $ do
      let readme = File "readme.txt" 1024 ".txt"
          config = File "config.json" 512 ".json"
          src = Directory "src" 
            [ File "main.fs" 2048 ".fs"
            , File "util.fs" 1024 ".fs"
            ]
          root = Directory "project" [readme, config, src]
          fsFiles = findByExtension ".fs" root
      length fsFiles `shouldBe` 2

    it "should get all files" $ do
      let readme = File "readme.txt" 1024 ".txt"
          config = File "config.json" 512 ".json"
          src = Directory "src" 
            [ File "main.fs" 2048 ".fs"
            , File "util.fs" 1024 ".fs"
            ]
          root = Directory "project" [readme, config, src]
          files = allFiles root
      length files `shouldBe` 4

    it "should add entries" $ do
      let dir = emptyDirectory "project"
          file = File "readme.txt" 1024 ".txt"
          updated = addEntry file dir
      fileCount updated `shouldBe` 1

-- ============================================================
-- MenuItem Tests
-- ============================================================

menuItemSpec :: Spec
menuItemSpec = describe "MenuItem" $ do
  describe "Item" $ do
    it "should have correct price" $ do
      let burger = Item "Hamburger" 500 "Main"
      menuPrice burger `shouldBe` 500

    it "should count as 1 item" $ do
      let burger = Item "Hamburger" 500 "Main"
      menuItemCount burger `shouldBe` 1

  describe "SetMenu" $ do
    it "should apply discount" $ do
      let burger = Item "Hamburger" 500 "Main"
          fries = Item "Fries" 200 "Side"
          drink = Item "Cola" 150 "Drink"
          setMenu = SetMenu "Lunch Set" [burger, fries, drink] 0.1
      -- Total: 850, with 10% discount: 765
      menuPrice setMenu `shouldBe` 765

    it "should count all items" $ do
      let burger = Item "Hamburger" 500 "Main"
          fries = Item "Fries" 200 "Side"
          drink = Item "Cola" 150 "Drink"
          setMenu = SetMenu "Lunch Set" [burger, fries, drink] 0.1
      menuItemCount setMenu `shouldBe` 3

    it "should create empty set menu" $ do
      let setMenu = emptySetMenu "Special" 0.2
      menuItemCount setMenu `shouldBe` 0

    it "should add items to set menu" $ do
      let setMenu = emptySetMenu "Special" 0.0
          burger = Item "Hamburger" 500 "Main"
          updated = addMenuItem burger setMenu
      menuItemCount updated `shouldBe` 1
      menuPrice updated `shouldBe` 500

-- ============================================================
-- Expression Tests
-- ============================================================

expressionSpec :: Spec
expressionSpec = describe "Expression" $ do
  describe "Number" $ do
    it "should evaluate to its value" $ do
      evaluate (Number 42) `shouldBe` Just 42

  describe "Variable" $ do
    it "should evaluate bound variable" $ do
      evaluate (Variable "x" (Just 10)) `shouldBe` Just 10

    it "should fail for unbound variable" $ do
      evaluate (Variable "x" Nothing) `shouldBe` Nothing

  describe "Add" $ do
    it "should add two numbers" $ do
      evaluate (Add (Number 2) (Number 3)) `shouldBe` Just 5

    it "should simplify x + 0 to x" $ do
      let expr = Add (Variable "x" (Just 5)) (Number 0)
      simplify expr `shouldBe` Variable "x" (Just 5)

    it "should simplify 0 + x to x" $ do
      let expr = Add (Number 0) (Variable "x" (Just 5))
      simplify expr `shouldBe` Variable "x" (Just 5)

    it "should simplify constant addition" $ do
      simplify (Add (Number 2) (Number 3)) `shouldBe` Number 5

  describe "Subtract" $ do
    it "should subtract two numbers" $ do
      evaluate (Subtract (Number 5) (Number 3)) `shouldBe` Just 2

    it "should simplify x - 0 to x" $ do
      simplify (Subtract (Variable "x" (Just 5)) (Number 0)) 
        `shouldBe` Variable "x" (Just 5)

    it "should simplify x - x to 0" $ do
      let x = Variable "x" Nothing
      simplify (Subtract x x) `shouldBe` Number 0

  describe "Multiply" $ do
    it "should multiply two numbers" $ do
      evaluate (Multiply (Number 4) (Number 5)) `shouldBe` Just 20

    it "should simplify x * 0 to 0" $ do
      simplify (Multiply (Variable "x" (Just 5)) (Number 0)) `shouldBe` Number 0

    it "should simplify 0 * x to 0" $ do
      simplify (Multiply (Number 0) (Variable "x" (Just 5))) `shouldBe` Number 0

    it "should simplify x * 1 to x" $ do
      simplify (Multiply (Variable "x" (Just 5)) (Number 1)) 
        `shouldBe` Variable "x" (Just 5)

    it "should simplify 1 * x to x" $ do
      simplify (Multiply (Number 1) (Variable "x" (Just 5))) 
        `shouldBe` Variable "x" (Just 5)

  describe "Divide" $ do
    it "should divide two numbers" $ do
      evaluate (Divide (Number 10) (Number 2)) `shouldBe` Just 5

    it "should return Nothing for division by zero" $ do
      evaluate (Divide (Number 10) (Number 0)) `shouldBe` Nothing

    it "should simplify 0 / x to 0" $ do
      simplify (Divide (Number 0) (Variable "x" (Just 5))) `shouldBe` Number 0

    it "should simplify x / 1 to x" $ do
      simplify (Divide (Variable "x" (Just 5)) (Number 1)) 
        `shouldBe` Variable "x" (Just 5)

  describe "Negate" $ do
    it "should negate a number" $ do
      evaluate (Negate (Number 5)) `shouldBe` Just (-5)

    it "should simplify double negation" $ do
      simplify (Negate (Negate (Variable "x" Nothing))) 
        `shouldBe` Variable "x" Nothing

  describe "Complex expressions" $ do
    it "should evaluate (2 + 3) * 4" $ do
      let expr = Multiply (Add (Number 2) (Number 3)) (Number 4)
      evaluate expr `shouldBe` Just 20

    it "should get variables" $ do
      let expr = Add (Variable "x" Nothing) (Variable "y" Nothing)
      variables expr `shouldBe` ["x", "y"]

    it "should bind variables" $ do
      let expr = Add (Variable "x" Nothing) (Variable "y" Nothing)
          bound = bindVars [("x", 10), ("y", 20)] expr
      evaluate bound `shouldBe` Just 30

-- ============================================================
-- Organization Tests
-- ============================================================

organizationSpec :: Spec
organizationSpec = describe "Organization" $ do
  describe "Employee" $ do
    it "should have correct salary" $ do
      let emp = Employee "Alice" 50000 "Engineer"
      totalSalary emp `shouldBe` 50000

    it "should count as 1 employee" $ do
      let emp = Employee "Alice" 50000 "Engineer"
      employeeCount emp `shouldBe` 1

  describe "Department" $ do
    it "should calculate total salary" $ do
      let alice = Employee "Alice" 50000 "Engineer"
          bob = Employee "Bob" 60000 "Engineer"
          dept = Department "Engineering" (Just "Charlie") [alice, bob]
      totalSalary dept `shouldBe` 110000

    it "should count all employees" $ do
      let alice = Employee "Alice" 50000 "Engineer"
          bob = Employee "Bob" 60000 "Engineer"
          charlie = Employee "Charlie" 70000 "Manager"
          eng = Department "Engineering" (Just "Charlie") [alice, bob]
          company = Department "Company" Nothing [eng, charlie]
      employeeCount company `shouldBe` 3

    it "should find employees by role" $ do
      let alice = Employee "Alice" 50000 "Engineer"
          bob = Employee "Bob" 60000 "Engineer"
          charlie = Employee "Charlie" 70000 "Manager"
          eng = Department "Engineering" (Just "Charlie") [alice, bob]
          company = Department "Company" Nothing [eng, charlie]
          engineers = findByRole "Engineer" company
      length engineers `shouldBe` 2

    it "should create empty department" $ do
      let dept = emptyDepartment "Sales" (Just "Dave")
      employeeCount dept `shouldBe` 0

    it "should add members" $ do
      let dept = emptyDepartment "Sales" Nothing
          emp = Employee "Eve" 45000 "Sales"
          updated = addMember emp dept
      employeeCount updated `shouldBe` 1

File: app/scala/part3/src/test/scala/CompositePatternSpec.scala
/**
 * 第7章: Composite パターン - テストコード
 */
package compositepattern

import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers

class CompositePatternSpec extends AnyFunSuite with Matchers:

  // ============================================
  // 1. Shape のテスト
  // ============================================
  
  test("Circle: translateは中心を移動する") {
    val circle = Circle(Point(10, 10), 5)
    val moved = circle.translate(3, 4)
    moved.center shouldBe Point(13, 14)
    moved.radius shouldBe 5
  }
  
  test("Circle: scaleは半径を拡大/縮小する") {
    val circle = Circle(Point(10, 10), 5)
    val scaled = circle.scale(2)
    scaled.radius shouldBe 10
    scaled.center shouldBe Point(10, 10)
  }
  
  test("Circle: areaは面積を計算する") {
    val circle = Circle(Point(0, 0), 2)
    circle.area should be ((math.Pi * 4) +- 0.001)
  }
  
  test("Circle: boundingBoxはバウンディングボックスを返す") {
    val circle = Circle(Point(10, 10), 5)
    val bb = circle.boundingBox
    bb.topLeft shouldBe Point(5, 5)
    bb.bottomRight shouldBe Point(15, 15)
  }
  
  test("Square: translateは左上を移動する") {
    val square = Square(Point(0, 0), 10)
    val moved = square.translate(5, 5)
    moved.topLeft shouldBe Point(5, 5)
    moved.side shouldBe 10
  }
  
  test("Square: scaleは辺を拡大/縮小する") {
    val square = Square(Point(0, 0), 10)
    val scaled = square.scale(2)
    scaled.side shouldBe 20
  }
  
  test("Square: areaは面積を計算する") {
    val square = Square(Point(0, 0), 5)
    square.area shouldBe 25
  }
  
  test("Rectangle: translateは左上を移動する") {
    val rect = Rectangle(Point(0, 0), 10, 5)
    val moved = rect.translate(3, 4)
    moved.topLeft shouldBe Point(3, 4)
  }
  
  test("Rectangle: scaleは幅と高さを拡大/縮小する") {
    val rect = Rectangle(Point(0, 0), 10, 5)
    val scaled = rect.scale(2)
    scaled.width shouldBe 20
    scaled.height shouldBe 10
  }
  
  test("Rectangle: areaは面積を計算する") {
    val rect = Rectangle(Point(0, 0), 10, 5)
    rect.area shouldBe 50
  }
  
  test("Triangle: areaは面積を計算する") {
    val triangle = Triangle(Point(0, 0), Point(4, 0), Point(2, 3))
    triangle.area should be (6.0 +- 0.001)
  }
  
  test("Triangle: translateは全ての頂点を移動する") {
    val triangle = Triangle(Point(0, 0), Point(4, 0), Point(2, 3))
    val moved = triangle.translate(1, 1)
    moved.p1 shouldBe Point(1, 1)
    moved.p2 shouldBe Point(5, 1)
    moved.p3 shouldBe Point(3, 4)
  }
  
  test("CompositeShape: addは図形を追加する") {
    val composite = CompositeShape()
      .add(Circle(Point(0, 0), 5))
      .add(Square(Point(10, 10), 5))
    composite.count shouldBe 2
  }
  
  test("CompositeShape: removeは図形を削除する") {
    val circle = Circle(Point(0, 0), 5)
    val square = Square(Point(10, 10), 5)
    val composite = CompositeShape()
      .add(circle)
      .add(square)
      .remove(circle)
    composite.count shouldBe 1
  }
  
  test("CompositeShape: translateは全ての子要素を移動する") {
    val composite = CompositeShape()
      .add(Circle(Point(0, 0), 5))
      .add(Square(Point(10, 10), 5))
    val moved = composite.translate(5, 5)
    
    moved.shapes(0).asInstanceOf[Circle].center shouldBe Point(5, 5)
    moved.shapes(1).asInstanceOf[Square].topLeft shouldBe Point(15, 15)
  }
  
  test("CompositeShape: scaleは全ての子要素を拡大/縮小する") {
    val composite = CompositeShape()
      .add(Circle(Point(0, 0), 5))
      .add(Square(Point(10, 10), 10))
    val scaled = composite.scale(2)
    
    scaled.shapes(0).asInstanceOf[Circle].radius shouldBe 10
    scaled.shapes(1).asInstanceOf[Square].side shouldBe 20
  }
  
  test("CompositeShape: areaは全ての子要素の面積の合計を返す") {
    val composite = CompositeShape()
      .add(Square(Point(0, 0), 10))  // 100
      .add(Rectangle(Point(0, 0), 5, 4))  // 20
    composite.area shouldBe 120
  }
  
  test("CompositeShape: boundingBoxは全ての子要素を含むバウンディングボックスを返す") {
    val composite = CompositeShape()
      .add(Circle(Point(0, 0), 5))   // (-5,-5) to (5,5)
      .add(Square(Point(10, 10), 5)) // (10,10) to (15,15)
    val bb = composite.boundingBox
    bb.topLeft shouldBe Point(-5, -5)
    bb.bottomRight shouldBe Point(15, 15)
  }
  
  test("CompositeShape: ネストした複合図形をサポートする") {
    val inner = CompositeShape()
      .add(Circle(Point(0, 0), 5))
    val outer = CompositeShape()
      .add(inner)
      .add(Square(Point(10, 10), 5))
    
    outer.count shouldBe 2
    outer.flatten.length shouldBe 2
  }
  
  test("CompositeShape: flattenはネストした図形をフラット化する") {
    val inner = CompositeShape()
      .add(Circle(Point(0, 0), 5))
      .add(Circle(Point(10, 10), 3))
    val outer = CompositeShape()
      .add(inner)
      .add(Square(Point(20, 20), 5))
    
    outer.flatten.length shouldBe 3
    outer.flatten.count(_.isInstanceOf[Circle]) shouldBe 2
    outer.flatten.count(_.isInstanceOf[Square]) shouldBe 1
  }

  // ============================================
  // 2. Switchable のテスト
  // ============================================
  
  test("Light: turnOnはオンにする") {
    val light = Light()
    val on = light.turnOn
    on.isOn shouldBe true
    on.on shouldBe true
  }
  
  test("Light: turnOffはオフにする") {
    val light = Light(on = true)
    val off = light.turnOff
    off.isOn shouldBe false
    off.on shouldBe false
  }
  
  test("DimmableLight: turnOnは明るさを100にする") {
    val light = DimmableLight()
    val on = light.turnOn
    on.intensity shouldBe 100
    on.isOn shouldBe true
  }
  
  test("DimmableLight: turnOffは明るさを0にする") {
    val light = DimmableLight(intensity = 50)
    val off = light.turnOff
    off.intensity shouldBe 0
    off.isOn shouldBe false
  }
  
  test("DimmableLight: setIntensityは明るさを設定する") {
    val light = DimmableLight()
    light.setIntensity(50).intensity shouldBe 50
    light.setIntensity(150).intensity shouldBe 100  // 上限
    light.setIntensity(-10).intensity shouldBe 0   // 下限
  }
  
  test("Fan: turnOnは中速でオンにする") {
    val fan = Fan()
    val on = fan.turnOn
    on.speed shouldBe 3
    on.isOn shouldBe true
  }
  
  test("Fan: setSpeedは速度を設定する") {
    val fan = Fan()
    fan.setSpeed(5).speed shouldBe 5
    fan.setSpeed(10).speed shouldBe 5  // 上限
    fan.setSpeed(-1).speed shouldBe 0  // 下限
  }
  
  test("CompositeSwitchable: addはスイッチを追加する") {
    val composite = CompositeSwitchable()
      .add(Light())
      .add(Fan())
    composite.count shouldBe 2
  }
  
  test("CompositeSwitchable: turnOnは全ての子要素をオンにする") {
    val composite = CompositeSwitchable()
      .add(Light())
      .add(DimmableLight())
      .add(Fan())
    val on = composite.turnOn
    
    on.allOn shouldBe true
    on.switchables.forall(_.isOn) shouldBe true
  }
  
  test("CompositeSwitchable: turnOffは全ての子要素をオフにする") {
    val composite = CompositeSwitchable()
      .add(Light(on = true))
      .add(DimmableLight(intensity = 100))
      .add(Fan(speed = 3))
    val off = composite.turnOff
    
    off.isOn shouldBe false
    off.switchables.forall(!_.isOn) shouldBe true
  }
  
  test("CompositeSwitchable: isOnは一つでもオンならtrue") {
    val composite = CompositeSwitchable()
      .add(Light(on = true))
      .add(Fan(speed = 0))
    composite.isOn shouldBe true
  }
  
  test("CompositeSwitchable: allOnは全てオンならtrue") {
    val allOn = CompositeSwitchable()
      .add(Light(on = true))
      .add(Fan(speed = 3))
    allOn.allOn shouldBe true
    
    val notAllOn = CompositeSwitchable()
      .add(Light(on = true))
      .add(Fan(speed = 0))
    notAllOn.allOn shouldBe false
  }
  
  test("CompositeSwitchable: ネストした複合スイッチをサポートする") {
    val bedroom = CompositeSwitchable(name = "Bedroom")
      .add(Light(name = "Ceiling"))
      .add(DimmableLight(name = "Bedside"))
    
    val livingRoom = CompositeSwitchable(name = "LivingRoom")
      .add(Light(name = "Main"))
      .add(Fan(name = "Ceiling Fan"))
    
    val house = CompositeSwitchable(name = "House")
      .add(bedroom)
      .add(livingRoom)
    
    val allOn = house.turnOn
    allOn.flatten.forall(_.isOn) shouldBe true
  }

  // ============================================
  // 3. FileSystem のテスト
  // ============================================
  
  test("File: sizeはファイルサイズを返す") {
    val file = File("test.txt", 100)
    file.size shouldBe 100
  }
  
  test("File: pathはファイルパスを返す") {
    val file = File("test.txt", 100, "home/user")
    file.path shouldBe "home/user/test.txt"
  }
  
  test("Directory: addはエントリを追加する") {
    val dir = Directory("root")
      .add(File("file1.txt", 100))
      .add(File("file2.txt", 200))
    dir.children.length shouldBe 2
  }
  
  test("Directory: sizeは全エントリのサイズの合計を返す") {
    val dir = Directory("root")
      .add(File("file1.txt", 100))
      .add(File("file2.txt", 200))
    dir.size shouldBe 300
  }
  
  test("Directory: ネストしたディレクトリのサイズを正しく計算する") {
    val subDir = Directory("subdir")
      .add(File("sub.txt", 50))
    
    val rootDir = Directory("root")
      .add(File("root.txt", 100))
      .add(subDir)
    
    rootDir.size shouldBe 150
  }
  
  test("Directory: fileCountはファイル数を返す") {
    val subDir = Directory("subdir")
      .add(File("sub1.txt", 50))
      .add(File("sub2.txt", 50))
    
    val rootDir = Directory("root")
      .add(File("root.txt", 100))
      .add(subDir)
    
    rootDir.fileCount shouldBe 3
  }
  
  test("Directory: directoryCountはディレクトリ数を返す") {
    val subDir1 = Directory("subdir1")
    val subDir2 = Directory("subdir2")
      .add(Directory("nested"))
    
    val rootDir = Directory("root")
      .add(subDir1)
      .add(subDir2)
    
    rootDir.directoryCount shouldBe 3
  }
  
  test("Directory: findは条件に一致するエントリを返す") {
    val dir = Directory("root")
      .add(File("test.txt", 100))
      .add(File("data.csv", 200))
      .add(File("test.csv", 150))
    
    val txtFiles = dir.find(e => e.name.endsWith(".txt"))
    txtFiles.length shouldBe 1
    
    val csvFiles = dir.find(e => e.name.endsWith(".csv"))
    csvFiles.length shouldBe 2
  }
  
  test("Directory: listAllは全てのパスをリストする") {
    val rootDir = Directory("root")
      .add(File("root.txt", 100))
      .add(Directory("subdir")
        .add(File("nested.txt", 50)))
    
    val paths = rootDir.listAll
    paths should contain("root/root.txt")
    paths should contain("root/subdir")
    // ネストしたファイルのパスも確認
    paths.exists(_.contains("nested.txt")) shouldBe true
  }

  // ============================================
  // 4. Menu のテスト
  // ============================================
  
  test("SingleItem: priceは価格を返す") {
    val item = SingleItem("Burger", BigDecimal(500), false, "Beef burger")
    item.price shouldBe BigDecimal(500)
  }
  
  test("SetMenu: priceは割引後の合計価格を返す") {
    val set = SetMenu("Lunch Set", discount = BigDecimal(100))
      .add(SingleItem("Burger", BigDecimal(500), false, ""))
      .add(SingleItem("Drink", BigDecimal(200), true, ""))
    
    set.originalPrice shouldBe BigDecimal(700)
    set.price shouldBe BigDecimal(600)
  }
  
  test("SetMenu: isVegetarianは全てベジタリアンならtrue") {
    val vegSet = SetMenu("Veg Set")
      .add(SingleItem("Salad", BigDecimal(300), true, ""))
      .add(SingleItem("Juice", BigDecimal(200), true, ""))
    vegSet.isVegetarian shouldBe true
    
    val mixSet = SetMenu("Mix Set")
      .add(SingleItem("Salad", BigDecimal(300), true, ""))
      .add(SingleItem("Burger", BigDecimal(500), false, ""))
    mixSet.isVegetarian shouldBe false
  }
  
  test("SetMenu: discountRateは割引率を返す") {
    val set = SetMenu("Set", discount = BigDecimal(100))
      .add(SingleItem("Item1", BigDecimal(500), false, ""))
      .add(SingleItem("Item2", BigDecimal(500), false, ""))
    
    set.discountRate shouldBe 10.0 +- 0.01
  }
  
  test("SetMenu: ネストしたセットメニューをサポートする") {
    val innerSet = SetMenu("Inner Set", discount = BigDecimal(50))
      .add(SingleItem("Item1", BigDecimal(300), false, ""))
      .add(SingleItem("Item2", BigDecimal(200), false, ""))
    
    val outerSet = SetMenu("Outer Set", discount = BigDecimal(100))
      .add(innerSet)
      .add(SingleItem("Item3", BigDecimal(400), false, ""))
    
    // Inner: 500 - 50 = 450, Outer: 450 + 400 - 100 = 750
    outerSet.price shouldBe BigDecimal(750)
  }

  // ============================================
  // 5. Expression のテスト
  // ============================================
  
  test("Number: evaluateは値を返す") {
    Number(42).evaluate shouldBe 42.0
  }
  
  test("Variable: evaluateは値を返す（バインドされている場合）") {
    Variable("x", Some(10)).evaluate shouldBe 10.0
  }
  
  test("Variable: evaluateは例外をスローする（バインドされていない場合）") {
    assertThrows[IllegalStateException] {
      Variable("x").evaluate
    }
  }
  
  test("Add: evaluateは加算結果を返す") {
    Add(Number(2), Number(3)).evaluate shouldBe 5.0
  }
  
  test("Subtract: evaluateは減算結果を返す") {
    Subtract(Number(5), Number(3)).evaluate shouldBe 2.0
  }
  
  test("Multiply: evaluateは乗算結果を返す") {
    Multiply(Number(4), Number(3)).evaluate shouldBe 12.0
  }
  
  test("Divide: evaluateは除算結果を返す") {
    Divide(Number(10), Number(2)).evaluate shouldBe 5.0
  }
  
  test("Divide: evaluateはゼロ除算で例外をスローする") {
    assertThrows[ArithmeticException] {
      Divide(Number(10), Number(0)).evaluate
    }
  }
  
  test("Expression: 複合式を評価できる") {
    // (2 + 3) * 4 = 20
    val expr = Multiply(Add(Number(2), Number(3)), Number(4))
    expr.evaluate shouldBe 20.0
  }
  
  test("Expression: variablesは未束縛変数を返す") {
    val expr = Add(Variable("x"), Multiply(Variable("y"), Number(2)))
    expr.variables shouldBe Set("x", "y")
  }
  
  test("Expression: bindは変数に値を束縛する") {
    val expr = Add(Variable("x"), Variable("y"))
    val bound = Expression.bind(expr, Map("x" -> 10.0, "y" -> 20.0))
    bound.evaluate shouldBe 30.0
  }
  
  test("Expression: simplifyは式を簡略化する") {
    // x + 0 = x
    Add(Variable("x"), Number(0)).simplify shouldBe Variable("x")
    
    // 0 + x = x
    Add(Number(0), Variable("x")).simplify shouldBe Variable("x")
    
    // x * 1 = x
    Multiply(Variable("x"), Number(1)).simplify shouldBe Variable("x")
    
    // x * 0 = 0
    Multiply(Variable("x"), Number(0)).simplify shouldBe Number(0)
    
    // 2 + 3 = 5
    Add(Number(2), Number(3)).simplify shouldBe Number(5)
  }
  
  test("Expression: simplifyは再帰的に簡略化する") {
    // (x + 0) * 1 = x
    val expr = Multiply(Add(Variable("x"), Number(0)), Number(1))
    expr.simplify shouldBe Variable("x")
  }
  
  test("Expression: toStringは式を文字列化する") {
    val expr = Add(Number(2), Multiply(Number(3), Number(4)))
    expr.toString shouldBe "(2.0 + (3.0 * 4.0))"
  }
